### PR TITLE
Privacy Widget (& Started with block)

### PR DESCRIPTION
**Pull request by JohnnyMcWeed**
Initial PR for privacy

Issue https://github.com/luyadev/luya-privacy/issues/1

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        "twbs/bootstrap": "v4.0.0-beta"
+	},
+	"autoload" : {
+		"psr-4" : {
+			"luya\\privacy\\" : "src/"
+		}
+	},
+	"autoload-dev" : {
+       "psr-4" : {
+           "privacytests\\" : "tests/"
```
Comment (nadar) [inline composer.json]: tests namespace should be `luya\\privacy\\tests`

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    },
+	"extra": {
+        "branch-alias": {
+            "dev-master": "1.0.x-dev"
+        },
+        "luya" : {
+        	"blocks" : ["src/frontend/blocks"]
+        }
+    },
+    "config": {
```
Comment (nadar) [inline composer.json]: I think we dont need this here.

File: composer.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+			"email": "alex.schmid@stud.unibas.ch"
+		}
+	],
+	"support" : {
+		"issues" : "https://github.com/luyadev/luya-privacy/issues"
+	},
+	"require-dev" : {
+	    "luyadev/luya-testsuite" : "~1.0.0",
+	    "php-coveralls/php-coveralls": "^1.0",
+        "twbs/bootstrap": "v4.0.0-beta"
```
Comment (nadar) [inline composer.json]: Why do wee need bootstrap and coveralls?

File: composer.json [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            "dev-master": "1.0.x-dev"
+        },
+        "luya" : {
+        	"blocks" : ["src/frontend/blocks"]
+        }
+    },
+    "config": {
+         "platform": {"php": "7.0"}  
+    },
+    "repositories": [
```
Comment (nadar) [inline composer.json]: Why do we need this here?

File: src/BasePrivacyWidget.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+use luya\privacy\assets\PrivacyWidgetAsset;
+use luya\privacy\traits\PrivacyTrait;
+
+/**
+ * Privacy Widget
+ *
+ * @author Alex Schmid <alex.schmid@stud.unibas.ch>
+ * @since 1.0.0
+ */
+class BasePrivacyWidget extends Widget
```
Comment (nadar) [inline src/BasePrivacyWidget.php]: Why using a base privacy widget? I would keep it clean and easier without an extra base class for the PrivacyWidget.

File: src/BasePrivacyBlock.php [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * @since 1.0.0
+ */
+abstract class BasePrivacyBlock extends PhpBlock
+{
+    /**
+     * @inheritdoc
+     */
+    public function getViewPath()
+    {
+        return '@privacy/views/blocks';
```
Comment (nadar) [inline src/BasePrivacyBlock.php]: We should not work with the fix module id, instead use this:

https://github.com/luyadev/luya-bootstrap3/blob/master/src/BaseBootstrap3Block.php#L19

File: src/assets/PrivacyAsset.php [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+     *   This functionality is available since version 2.0.7.
+     *
+     * Note that only a forward slash "/" should be used as directory separator.
+     */
+    public $jsOnPrivacyAccepted = [];
+
+    /**
+     * @inheritdoc
+     */
+    function registerAssetFiles($view)
```
Comment (nadar) [inline src/assets/PrivacyAsset.php]: This looks very complex to me. Why not just do it like this:

```php
public function init()
{
     parent::init();
    if ($this->getIsOnPrivacyAccepte()) {
    $this->js = array_merge($this->js, $this->jsOnPrivacyAccepted);
    }
}
```

File: src/traits/PrivacyTrait.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+ * 
+ * @author Alex Schmid <alex.schmid@stud.unibas.ch>
+ * @since 1.0.0
+ */
+trait PrivacyTrait
+{
+    /**
+     * @return bool|null The privacy value
+     */
+    public function getPrivacy()
```
Comment (nadar) [inline src/traits/PrivacyTrait.php]: I would suggest to use another function name, which really says what it does:

getPrivacyCookieValue

File: src/traits/PrivacyTrait.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        return Yii::$app->response->cookies->getValue('_privacyPolicy', null) !== null ?
+            Yii::$app->response->cookies->getValue('_privacyPolicy') : 
+            Yii::$app->request->cookies->getValue('_privacyPolicy', null);
+    }
+
+    /**
+     * Sets the privacy
+     */
+    public function setPrivacy($value = null)
```
Comment (nadar) [inline src/traits/PrivacyTrait.php]: Same here: setPrivacyCookieValue

File: src/traits/PrivacyTrait.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Trait PrivacyTrait
+ * 
+ * The privacy trait helps getting and setting the privacy policy state.
+ * 
+ * @author Alex Schmid <alex.schmid@stud.unibas.ch>
+ * @since 1.0.0
+ */
+trait PrivacyTrait
```
Comment (nadar) [inline src/traits/PrivacyTrait.php]: use a CONST for the cookie name defintion like

const PRIVACY_COOKIE_NAME = '_privacyPolicy'

File: src/widgets/PrivacyWidget.php [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        $this->setMessages();
+        Module::onLoad();
```
Comment (nadar) [inline src/widgets/PrivacyWidget.php]: this is not needed - i you run Module::t - the onLoad will be triggered.

File: src/widgets/PrivacyWidget.php [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        $privacyPolicy = Yii::$app->request->post('_privacyPolicy', null);
+        if (!empty($privacyPolicy)){
+            if ($privacyPolicy === 'true') {
+                Yii::$app->response->cookies->add(new Cookie(["name" => "_privacyPolicy", "value" => true]));
```
Comment (nadar) [inline src/widgets/PrivacyWidget.php]: Why arent you using the funtions from the trait ... this is why we have the trait - if am right.

File: src/widgets/PrivacyWidget.php [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        $this->setMessages();
+        Module::onLoad();
+        $privacyPolicy = Yii::$app->request->post('_privacyPolicy', null);
+        if (!empty($privacyPolicy)){
+            if ($privacyPolicy === 'true') {
```
Comment (nadar) [inline src/widgets/PrivacyWidget.php]: Maybe we should use an int value instead of string comparing.

**Review by nadar — CHANGES_REQUESTED** [review]
Thanks, please take a look at my review.

File: src/traits/PrivacyTrait.php (hunk 10 above)
Comment (JohnnyMcWeed) [reply src/traits/PrivacyTrait.php]: Not possible in trait.. So where do you want that?
Changed it to static for the moment..

File: composer.json [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+       "psr-4" : {
+           "luya\\privacy\\tests\\" : "tests/"
+       }
+    },
+	"extra": {
+        "luya" : {
+        	"blocks" : ["src/frontend/blocks"]
+        }
+    },
+    "repositories": [
```
Comment (nadar) [inline composer.json]: @JohnnyMcWeed why do wee need this one?

File: src/traits/PrivacyTrait.php (hunk 10 above)
Comment (nadar) [reply src/traits/PrivacyTrait.php]: ok

File: src/widgets/PrivacyWidget.php [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+ * @since 1.0.0
+ * 
+ * @todo Widget id must be better
+ * @todo Position -> fixed / relative
+ */
+class PrivacyWidget extends Widget
+{
+    use PrivacyTrait;
+
+    const PRIVACY_COOKIE_NAME = '_privacyPolicy';
```
Comment (nadar) [inline src/widgets/PrivacyWidget.php]: Why should we repeat this? Its part of the PrivacyTrait.

File: src/widgets/PrivacyWidget.php [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+
+    /**
+     * @inheritdoc
+     */
+    public function getViewPath()
+    {
+        return '@privacy/views/widgets';
```
Comment (nadar) [inline src/widgets/PrivacyWidget.php]: You don't need to define a view path. This is should work out of the box for widgets.

File: src/widgets/PrivacyWidget.php [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        $this->setMessages();
+        $privacyPolicy = Yii::$app->request->post(self::PRIVACY_COOKIE_NAME, null);
+        if (!empty($privacyPolicy)){
+            if ($privacyPolicy === 'true') {
+                Yii::$app->response->cookies->add(new Cookie(["name" => self::PRIVACY_COOKIE_NAME, "value" => true]));
```
Comment (nadar) [inline src/widgets/PrivacyWidget.php]: @JohnnyMcWeed setting the cookie should be part of the trait and not the widget. As maybe also other implementations could trigger this (assuming a checkbox on user registration which also says "allow cookies" - now the controller could implement the trait)

Comment (nadar) [reply src/widgets/PrivacyWidget.php]: So all we need here is to catch the post request and run the PrivacyTrait setPrivacyCookie method.

Comment (nadar) [reply src/widgets/PrivacyWidget.php]: unless this cookie is not set - cookies are not allowed. No need to delete/removeAll.

File: src/widgets/PrivacyWidget.php [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+        }
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function run()
+    {
+        if ((!$this->isPrivacyAccepted() && !$this->isPrivacyDeclined()) || $this->forceOutput) {
+            if (!$this->removeCSS) $this->getView()->registerCss($this->css);
```
Comment (nadar) [inline src/widgets/PrivacyWidget.php]: I don't think its a good practice to have $removeCass variable, if people want to remove the defautl css they can just override the css with `'css' => false`

File: tests/src/PrivacyTestCase.php [hunk 19]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace luya\privacytests;
```
Comment (nadar) [inline tests/src/PrivacyTestCase.php]: wrong namespaces.

Comment (nadar) [reply tests/src/PrivacyTestCase.php]: should be luya\privacy\tests

File: composer.json [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+		{
+			"name" : "Basil Suter",
+			"email" : "git@nadar.io",
+			"homepage" : "https://github.com/nadar"
+		},
+		{
+			"name": "Alex Schmid",
+			"email": "alex.schmid@stud.unibas.ch"
+		}
+	],
```
Comment (nadar) [inline composer.json:18]: require-dev section should include test suite, as you had it before.

**Review by nadar — CHANGES_REQUESTED** [review]

**Pull request merged**